### DISCUSSION TITLE
Lab6 Timing Violation

### DISCUSSION THREAD

**Discussion by HoraceKuo** [bol-edu/HLS-SOC-Discussions#158] [Q&A]
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/145552964/cc9c84d9-167d-4211-89e4-8ef22da8be2b)
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/145552964/3f59d371-0a75-481f-9d28-134465aef3a2)

在模擬測試過結合後的各功能正常，但跑合成時出現failed to meet the timing requirement

再來把合成後的bitstream放到fpga上能夠看到hello逐字印出，而checkbit卻還是讀到AB40，不是結束時的AB51

想請問如何解決timing violation的問題，而ipynb中所讀到的checkbit又為何會停在AB40而不是AB51?
下面附上timing report和ipynb執行結果

![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/145552964/a04a7e7f-48b1-4ae3-b021-150b6b385bb6)
<img width="725" alt="image" src="https://github.com/bol-edu/HLS-SOC-Discussions/assets/145552964/eabc4c33-f9c1-4acb-98f1-0531205e0f88">

**Comment by ghost** [comment]
幫隊友補充一下，後來我在notebook加入以下程式(下圖1)，試想再啟動uart前先看放在38000000的firmware有沒有運作，發現它會卡在ab40就不動了，但我們在模擬時firmware有完整做完，處理完irc後也有把後面firmware做完(下圖2)，不知道為何放在FPGA就動不了了。
![messageImage_1701442818642](https://github.com/bol-edu/HLS-SOC-Discussions/assets/119398292/5c472183-5da1-4c88-b9ab-e4cfc229bce3)
![messageImage_1701431998038](https://github.com/bol-edu/HLS-SOC-Discussions/assets/119398292/7ed8de19-d37d-4aa2-8425-89d325d3fe45)

**Comment by stage0831** [comment]
同學你好，關於failed to meet the timing requirement的問題，我這邊也有遇到。

我們嘗試進到vivado並改變implementation的strategy，我們將default改成使用performance_NetDelay_high後有成功修復我們的hold time violation，可以試試看這個方法，希望有點幫助。
![ref](https://github.com/bol-edu/HLS-SOC-Discussions/assets/72055234/30848baf-141e-4051-8d6f-6e2d54f9d101)

另外，想請教你們一個問題，如下圖波型

![ref1](https://github.com/bol-edu/HLS-SOC-Discussions/assets/72055234/357036ee-082f-40aa-9047-ef649b8dbd2d)

我們在最後燒入FPGA後並沒有收到hello這個字串。我們在tb端送workload的順序是MM->QS->FIR，然後UART會在MM結束時同時發出中斷，想請問你們在tb端送workload的順序是如何的?

**Reply by ghost** [reply]
非常感謝你的幫忙，我會在試看看。我在模擬時順序跟你們一樣，我有試過在MM結束後、MM開始前與fir結束三種狀況。如果FPGA收不到hello回傳，我猜可能是irq沒送到risc-v CPU，到risc-v CPU的irq我是只接到30000000而已，給你參考。
另外能請問你，收完hello你們firmware有繼續運行嗎?我們的會卡住(下圖所示)
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/119398292/17da784e-ffac-4e5b-bcff-c37c03becc02)

**Reply by BarryJu1999** [reply]
同學你好，想請教這個問題你們後來是如何解決的呢?
我們目前也遇到相同的問題，但不知道如何處理。

**Reply by stage0831** [reply]
我目前也是進行到這裡卡住，猜測是bram那邊的問題，因為到下圖指令的位置就會卡住。
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/72055234/61e5e75f-2f77-4508-b174-a9590a8f3871)
可以由python這邊看到卡住的位置。
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/72055234/5c05d774-24a3-4496-9e49-87cba1d30d7f)

**Reply by ghost** [reply]
同學你好，我們這邊目前也還沒解決

**Reply by LingChiYang** [reply]
同學好

助教這邊看到很多人在整合後都有firmware不會繼續執行的問題。同學可以注意以下細節

1. 在模擬完，整合後的hardware也需更新至vivado folder以便於fpga implement
https://github.com/bol-edu/caravel-soc_fpga-lab/tree/main/lab-wlos_baseline/vivado/vvd_srcs/caravel_soc/rtl/user
2. vivado folder底下tcl及include的file有可能也需要更新
https://github.com/bol-edu/caravel-soc_fpga-lab/blob/main/lab-wlos_baseline/vivado/vvd_caravel_fpga_40mhz.tcl#L30-L49
https://github.com/bol-edu/caravel-soc_fpga-lab/blob/main/lab-wlos_baseline/vivado/vvd_caravel_fpga_40mhz.tcl#L197-L216

**Reply by sonnyshiau** [reply]
同學您好，請問當時遇到timing violation最後是如何解決
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/126078155/10710f95-1124-44db-b647-878c0c2cc9e8)

![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/126078155/a8fa932f-b5c9-4a2e-9666-98c8c2622572)

**Reply by chickenchennnn** [reply]
我們也有遇到這個問題，解法上面有一位同學有提到，進到vivado並改變implementation的strategy，使用performance_NetDelay_high有成功解決我們hold time的violation

**Reply by sonnyshiau** [reply]
同學您好，有使用上面的方法但是還是會有hold time violation

**Reply by sonnyshiau** [reply]
附上jupyter的結果
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/126078155/d786e199-38c6-4036-8a6d-31dc2e067c8b)

**Reply by ZGCiou** [reply]
可以設定phys_opt_design參數來修復hold time violation
<img width="884" alt="截圖 2023-12-12 下午3 52 52" src="https://github.com/bol-edu/HLS-SOC-Discussions/assets/127314907/a8354ea8-d670-4295-b385-ae05f83f37ba">
-derictive 選擇 ExploreWithHoldFix

或是使用tcl
在 launch_runs 前加入 set_property STEPS.PHYS_OPT_DESIGN.ARGS.DIRECTIVE ExploreWithHoldFix [get_runs impl_1]

**Reply by sonnyshiau** [reply]
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/126078155/ae476d24-c82b-4e95-a007-27f9b9ba947d)
我最後直接在tcl改strategy就改善了，感謝同學

**Comment by LingChiYang** [comment]
同學好

助教這邊跑是有繼續跑的，不太確定同學有做甚麼改動。我想有幾點需要提醒的是，放到FPGA時isr大概率會是在main的最後跑，因為很快就執行完了，所以同學可以試試看irq在main的最後拉起來模擬看看。

**Comment by Josephood7** [comment]
同學好~
可以請問同學你們的bram.v那邊一樣是設N = 10嗎?
因為我把firmware合起來以後連LA test 1都跑不出來就time out了

**Reply by huankai0731** [reply]
可以試試看把time out關掉，firmware合起來之後好像會run比較久

**Reply by Josephood7** [reply]
謝謝你的回覆~ 但跑了幾十分鐘還是沒有東西
你是把   firmware.c   firmware.h   tb.v   la.c  合起來對嗎
有其他檔案也需要動的ㄇ

**Reply by ghost** [reply]
我3800000那邊的設8K(N=11)，N=10太小會不過，你可以試看看
